# GBA 6070 - Programming Foundation for Business Analytics
# Dr. Mohammad Salehan
# Assignment 10 - Dataframe II
Enter your name below.

In this assignment you will work with a weather dataset. Let's start with loading the dataset.

In [55]:
import pandas as pd
import numpy as np
weather = pd.read_csv('weather2.csv')

We also use another dataset that contains information about each weather station.

In [56]:
stations = pd.read_excel('Station Information.xlsx')
stations.shape

(29745, 10)

1. Examine the shape of the both dataframes.

In [57]:
weather.shape, stations.shape

((100000, 14), (29745, 10))

2. Check the top 5 rows of each dataframe to see what it looks like.

In [58]:
weather.head()

Unnamed: 0  station_wban    year  month   day  hour  temperature  \
0           0  841170-99999  2010.0    1.0  15.0  16.0        270.0   
1           1  841170-99999  2010.0    1.0  15.0  17.0        270.0   
2           2  841170-99999  2010.0    1.0  15.0  18.0        276.0   
3           3  841170-99999  2010.0    1.0  15.0  19.0        280.0   
4           4  841170-99999  2010.0    1.0  15.0  20.0        280.0   

   dew_point  pressure  wind_direction  wind_speed  sky_condition  \
0      210.0   -9999.0           300.0        51.0            7.0   
1      210.0   -9999.0           310.0        57.0            7.0   
2      208.0   10115.0           310.0        51.0            7.0   
3      220.0   -9999.0           320.0        62.0            7.0   
4      220.0   -9999.0           300.0        72.0            7.0   

   precipitation_1h  precipitation_6h  
0           -9999.0           -9999.0  
1           -9999.0           -9999.0  
2           -9999.0           -9999.0  
3           -9999.0           -9999.0  
4           -9999.0           -9999.0

In [59]:
stations.head()

USAF   WBAN STATION NAME CTRY ST CALL    LAT     LON  ELEV(M)     BEGIN  \
0  7018  99999   WXPOD 7018  NaN     NaN   0.00   0.000   7018.0  20110309   
1  7026  99999   WXPOD 7026   AF     NaN   0.00   0.000   7026.0  20120713   
2  7070  99999   WXPOD 7070   AF     NaN   0.00   0.000   7070.0  20140923   
3  8260  99999    WXPOD8270  NaN     NaN   0.00   0.000      0.0  20050101   
4  8268  99999    WXPOD8278   AF     NaN  32.95  65.567   1156.7  20100519   

        END  
0  20130730  
1  20170822  
2  20150926  
3  20100920  
4  20120323

3. Drop the first column (Unnamed) of weather, then display the top row.

In [60]:
weather= weather.iloc[:, 1:]
weather.head(1)

station_wban    year  month   day  hour  temperature  dew_point  pressure  \
0  841170-99999  2010.0    1.0  15.0  16.0        270.0      210.0   -9999.0   

   wind_direction  wind_speed  sky_condition  precipitation_1h  \
0           300.0        51.0            7.0           -9999.0   

   precipitation_6h  
0           -9999.0

Below, we combine USAF and WBAN columns of stations dataframe to create a new column named station_wban that looks like the one in weather dataframe.

In [61]:
weather.shape

(100000, 13)

In [62]:
stations.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29745 entries, 0 to 29744
Data columns (total 10 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   USAF          29745 non-null  object 
 1   WBAN          29745 non-null  int64  
 2   STATION NAME  28955 non-null  object 
 3   CTRY          28811 non-null  object 
 4   ST CALL       12501 non-null  object 
 5   LAT           28546 non-null  float64
 6   LON           28545 non-null  float64
 7   ELEV(M)       28448 non-null  float64
 8   BEGIN         29745 non-null  int64  
 9   END           29745 non-null  int64  
dtypes: float64(3), int64(3), object(4)
memory usage: 2.3+ MB


In [63]:
stations['station_wban'] = stations['USAF'].astype(str) + '-' + stations['WBAN'].astype(str)
stations.head()

USAF   WBAN STATION NAME CTRY ST CALL    LAT     LON  ELEV(M)     BEGIN  \
0  7018  99999   WXPOD 7018  NaN     NaN   0.00   0.000   7018.0  20110309   
1  7026  99999   WXPOD 7026   AF     NaN   0.00   0.000   7026.0  20120713   
2  7070  99999   WXPOD 7070   AF     NaN   0.00   0.000   7070.0  20140923   
3  8260  99999    WXPOD8270  NaN     NaN   0.00   0.000      0.0  20050101   
4  8268  99999    WXPOD8278   AF     NaN  32.95  65.567   1156.7  20100519   

        END station_wban  
0  20130730   7018-99999  
1  20170822   7026-99999  
2  20150926   7070-99999  
3  20100920   8260-99999  
4  20120323   8268-99999

4. Join the 2 dataframes to add station name, country (CTRY) and state (ST CALL) to weather dataframe. There is no need for the rest of the columns in stations. You should use the proper type of join so that the number of rows in weather dataframe stays the same. At the end of this part, the shape of weather should be (100000, 16).

In [64]:
weather = pd.merge(weather, stations, how='left', on='station_wban')
weather.head(1)

station_wban    year  month   day  hour  temperature  dew_point  pressure  \
0  841170-99999  2010.0    1.0  15.0  16.0        270.0      210.0   -9999.0   

   wind_direction  wind_speed  ...    USAF   WBAN      STATION NAME CTRY  \
0           300.0        51.0  ...  841170  99999  ELOY ALFARO INTL   EC   

   ST CALL   LAT     LON ELEV(M)     BEGIN       END  
0     SEMT -0.95 -80.683    14.6  19770701  20200329  

[1 rows x 23 columns]

In [65]:
weather.drop(['USAF', 'WBAN', 'LAT', 'LON', 'ELEV(M)', 'BEGIN', 'END'], axis=1, inplace=True)
weather.head(1)

station_wban    year  month   day  hour  temperature  dew_point  pressure  \
0  841170-99999  2010.0    1.0  15.0  16.0        270.0      210.0   -9999.0   

   wind_direction  wind_speed  sky_condition  precipitation_1h  \
0           300.0        51.0            7.0           -9999.0   

   precipitation_6h      STATION NAME CTRY ST CALL  
0           -9999.0  ELOY ALFARO INTL   EC    SEMT

In [66]:
weather.shape

(100000, 16)

5. Show descriptive statistics for all numeric columns of weather dataframe.

In [67]:
weather.describe()

year          month            day           hour  \
count  100000.00000  100000.000000  100000.000000  100000.000000   
mean     2009.34663       6.751760      15.871240      11.483780   
std         5.49655       3.392016       8.831961       6.926196   
min      1999.00000       1.000000       1.000000       0.000000   
25%      2005.00000       4.000000       8.000000       5.000000   
50%      2009.00000       7.000000      16.000000      11.000000   
75%      2015.00000      10.000000      24.000000      17.000000   
max      2019.00000      12.000000      31.000000      23.000000   

         temperature     dew_point       pressure  wind_direction  \
count  100000.000000  100000.00000  100000.000000   100000.000000   
mean      136.975390      81.14740    5797.130830       23.626870   
std       361.774103     362.78918    8303.764083     1264.655425   
min     -9999.000000   -9999.00000   -9999.000000    -9999.000000   
25%        61.000000      13.00000   10037.000000       70.000000   
50%       156.000000      99.00000   10140.000000      180.000000   
75%       239.000000     178.00000   10189.000000      270.000000   
max      9999.000000    9999.00000   10449.000000      999.000000   

         wind_speed  sky_condition  precipitation_1h  precipitation_6h  
count  100000.00000  100000.000000      100000.00000     100000.000000  
mean       33.54303   -1839.430760       -4556.22822      -9511.991990  
std       110.06946    3878.528361        4980.57331       2154.103991  
min     -9999.00000   -9999.000000       -9999.00000      -9999.000000  
25%        20.00000       0.000000       -9999.00000      -9999.000000  
50%        31.00000       2.000000          -1.00000      -9999.000000  
75%        46.00000       7.000000           0.00000      -9999.000000  
max       999.00000       9.000000        2794.00000       1278.000000

6. In all numeric columns, values 9999, -9999, 999, and -999 represent missing values. Replace all of them with np.nan using <a href='https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.replace.html' target='blank'>.replace() </a> method of dataframe.

In [68]:
weather = weather.replace([9999, -9999, 999,  -999], np.nan)
weather.head()

station_wban    year  month   day  hour  temperature  dew_point  pressure  \
0  841170-99999  2010.0    1.0  15.0  16.0        270.0      210.0       NaN   
1  841170-99999  2010.0    1.0  15.0  17.0        270.0      210.0       NaN   
2  841170-99999  2010.0    1.0  15.0  18.0        276.0      208.0   10115.0   
3  841170-99999  2010.0    1.0  15.0  19.0        280.0      220.0       NaN   
4  841170-99999  2010.0    1.0  15.0  20.0        280.0      220.0       NaN   

   wind_direction  wind_speed  sky_condition  precipitation_1h  \
0           300.0        51.0            7.0               NaN   
1           310.0        57.0            7.0               NaN   
2           310.0        51.0            7.0               NaN   
3           320.0        62.0            7.0               NaN   
4           300.0        72.0            7.0               NaN   

   precipitation_6h      STATION NAME CTRY ST CALL  
0               NaN  ELOY ALFARO INTL   EC    SEMT  
1               NaN  ELOY ALFARO INTL   EC    SEMT  
2               NaN  ELOY ALFARO INTL   EC    SEMT  
3               NaN  ELOY ALFARO INTL   EC    SEMT  
4               NaN  ELOY ALFARO INTL   EC    SEMT

7. Run descriptive statistics again to make sure the the values are transformed correctly.

In [69]:
weather.describe()

year          month            day           hour  \
count  100000.00000  100000.000000  100000.000000  100000.000000   
mean     2009.34663       6.751760      15.871240      11.483780   
std         5.49655       3.392016       8.831961       6.926196   
min      1999.00000       1.000000       1.000000       0.000000   
25%      2005.00000       4.000000       8.000000       5.000000   
50%      2009.00000       7.000000      16.000000      11.000000   
75%      2015.00000      10.000000      24.000000      17.000000   
max      2019.00000      12.000000      31.000000      23.000000   

        temperature     dew_point      pressure  wind_direction    wind_speed  \
count  99711.000000  99698.000000  78317.000000    97501.000000  99937.000000   
mean     144.152882     87.462597  10162.052505      173.352971     34.144951   
std      110.650054    108.562320     71.561724      112.883343     23.637969   
min     -289.000000   -333.000000   9833.000000        0.000000      0.000000   
25%       61.000000     14.000000  10121.000000       70.000000     20.000000   
50%      156.000000     99.000000  10159.000000      180.000000     31.000000   
75%      239.000000    178.000000  10204.000000      270.000000     46.000000   
max      420.000000    280.000000  10449.000000      360.000000    175.000000   

       sky_condition  precipitation_1h  precipitation_6h  
count   81570.000000      54426.000000       4863.000000  
mean        4.149736          1.315621         15.559120  
std         3.341461         18.754512         50.480649  
min         0.000000         -1.000000         -1.000000  
25%         0.000000          0.000000          0.000000  
50%         4.000000          0.000000          0.000000  
75%         8.000000          0.000000          8.000000  
max         9.000000       2794.000000       1278.000000

8. All weather meaures are inflated with a factor of 10. For example, 289 for temperature means 28.9 degrees. Divide all temperature values by 10. Use .apply() method and lambda function. Other columns, such as day and hour should not be affected.

In [70]:
weather[['temperature','dew_point', 'pressure', 'wind_direction', 'wind_speed',
         'precipitation_1h', 'precipitation_6h']] = weather[[
    'temperature','dew_point', 'pressure', 'wind_direction', 'wind_speed', 
    'precipitation_1h', 'precipitation_6h']]/10
weather.head()

station_wban    year  month   day  hour  temperature  dew_point  pressure  \
0  841170-99999  2010.0    1.0  15.0  16.0         27.0       21.0       NaN   
1  841170-99999  2010.0    1.0  15.0  17.0         27.0       21.0       NaN   
2  841170-99999  2010.0    1.0  15.0  18.0         27.6       20.8    1011.5   
3  841170-99999  2010.0    1.0  15.0  19.0         28.0       22.0       NaN   
4  841170-99999  2010.0    1.0  15.0  20.0         28.0       22.0       NaN   

   wind_direction  wind_speed  sky_condition  precipitation_1h  \
0            30.0         5.1            7.0               NaN   
1            31.0         5.7            7.0               NaN   
2            31.0         5.1            7.0               NaN   
3            32.0         6.2            7.0               NaN   
4            30.0         7.2            7.0               NaN   

   precipitation_6h      STATION NAME CTRY ST CALL  
0               NaN  ELOY ALFARO INTL   EC    SEMT  
1               NaN  ELOY ALFARO INTL   EC    SEMT  
2               NaN  ELOY ALFARO INTL   EC    SEMT  
3               NaN  ELOY ALFARO INTL   EC    SEMT  
4               NaN  ELOY ALFARO INTL   EC    SEMT

9. Using a copy of your dataframe, replace all missing values with mean of each column.

In [71]:
weather2 = weather.fillna(weather.mean())
weather2.head()

C:\Users\msalehan\AppData\Local\Temp\1/ipykernel_10584/1760368082.py:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  weather2 = weather.fillna(weather.mean())


station_wban    year  month   day  hour  temperature  dew_point  \
0  841170-99999  2010.0    1.0  15.0  16.0         27.0       21.0   
1  841170-99999  2010.0    1.0  15.0  17.0         27.0       21.0   
2  841170-99999  2010.0    1.0  15.0  18.0         27.6       20.8   
3  841170-99999  2010.0    1.0  15.0  19.0         28.0       22.0   
4  841170-99999  2010.0    1.0  15.0  20.0         28.0       22.0   

     pressure  wind_direction  wind_speed  sky_condition  precipitation_1h  \
0  1016.20525            30.0         5.1            7.0          0.131562   
1  1016.20525            31.0         5.7            7.0          0.131562   
2  1011.50000            31.0         5.1            7.0          0.131562   
3  1016.20525            32.0         6.2            7.0          0.131562   
4  1016.20525            30.0         7.2            7.0          0.131562   

   precipitation_6h      STATION NAME CTRY ST CALL  
0          1.555912  ELOY ALFARO INTL   EC    SEMT  
1          1.555912  ELOY ALFARO INTL   EC    SEMT  
2          1.555912  ELOY ALFARO INTL   EC    SEMT  
3          1.555912  ELOY ALFARO INTL   EC    SEMT  
4          1.555912  ELOY ALFARO INTL   EC    SEMT

10. Calculate mean temperature, max wind speed, and min dew point for each station for each year using the dataframe from the previous step.

In [72]:
weather2.groupby(['station_wban', 'year']).aggregate({
    'temperature': 'mean',
    'dew_point': 'min',
    'wind_speed': 'max'
})

temperature  dew_point  wind_speed
station_wban year                                      
156550-99999 2007.0     6.693777   -9.10000        15.0
164590-99999 1999.0    12.272839   -3.00000        17.5
             2006.0    22.144752   -9.00000         9.8
167460-99999 2001.0    11.628309   -1.00000         9.9
383531-99999 2004.0     7.250646  -20.00000        11.0
486940-99999 2005.0    29.219368   21.00000         6.7
486980-99999 2001.0    26.770476    8.74626         8.7
623660-99999 2003.0    16.276334   -9.50000         9.9
             2006.0    12.415385    4.70000         5.1
             2015.0    18.624924   -4.00000         8.8
702730-26451 2013.0    16.026901    3.90000        10.8
722280-13876 1999.0    28.673373   16.80000         6.7
             2007.0    29.559903   10.00000         9.9
             2008.0     8.727381  -17.80000         9.8
             2011.0    16.524815   -7.20000        13.9
722436-12906 2013.0    29.141351    8.74626         9.9
722745-99999 2003.0    20.194030  -16.00000         9.3
722900-23188 2014.0    22.361448   -3.90000         7.7
723065-99999 2000.0    21.237470   -6.00000        12.9
724010-13740 2003.0    25.575140   11.70000         8.2
             2011.0    10.868392  -10.60000        13.9
724565-99999 2002.0     5.127315  -14.00000        12.3
724660-93037 2002.0     0.153400  -26.70000         9.8
724695-23036 2005.0    16.670402   -9.00000         9.8
             2009.0    15.676407  -12.00000        14.4
             2016.0     6.432739  -22.30000         9.8
725200-94823 2000.0     8.413071   -9.40000        13.9
             2006.0     4.635102   -7.80000         9.9
             2018.0    22.914980    8.90000        12.9
725335-94833 2016.0     6.551405   -9.90000        17.5
725460-14933 2002.0    -0.375000  -20.60000         9.8
725540-14949 2005.0    23.189887    9.00000         9.3
             2017.0    21.064949    0.00000         9.8
             2019.0    23.557562    8.00000         8.2
726088-14606 2006.0    -0.086576  -26.70000         9.7
726170-14742 2000.0     0.862165  -25.00000        12.9
             2008.0     3.005140  -24.40000        13.9
726980-24229 2009.0    16.973607   -9.40000         9.9
             2010.0    15.341126   -5.00000         9.9
726986-99999 2004.0    17.671759    1.00000         8.2
727530-14914 2016.0    12.650715  -17.20000        16.0
             2018.0    15.750718   -8.90000        15.4
727676-24013 2007.0    -3.680451  -33.30000        16.5
727850-24157 2010.0     5.284889   -9.40000        15.9
             2018.0    20.653906   -3.90000        10.8
742060-24207 2005.0    19.185776    1.00000         7.2
             2009.0    14.185795   -6.00000         9.8
             2012.0     4.693801   -7.10000        13.4
745700-13840 2011.0    14.094710    0.00000        15.9
745980-13702 2008.0    24.344397   11.60000         8.8
746710-13806 2009.0     5.929665  -15.50000         9.8
785350-11630 2011.0    26.928214    8.74626         9.9
841170-99999 2010.0    26.217236    8.74626         8.7
             2015.0    25.593533    8.74626         9.9
             2019.0    23.636676    8.74626         9.9
911820-22521 2015.0    22.723810    7.20000        14.4
912850-21504 2017.0    24.594244   12.20000         9.3

11. When imputing missing values, replace each missing values with the mean of the measure for the same station. Use the dataframe from step 8.

In [74]:
weather3 = weather.fillna(weather.groupby('station_wban').transform('mean'))
weather3.head()

C:\Users\msalehan\AppData\Local\Temp\1/ipykernel_10584/2996662162.py:1: FutureWarning: Dropping invalid columns in DataFrameGroupBy.transform is deprecated. In a future version, a TypeError will be raised. Before calling .transform, select only columns which should be valid for the transforming function.
  weather3 = weather.fillna(weather.groupby('station_wban').transform('mean'))


station_wban    year  month   day  hour  temperature  dew_point  \
0  841170-99999  2010.0    1.0  15.0  16.0         27.0       21.0   
1  841170-99999  2010.0    1.0  15.0  17.0         27.0       21.0   
2  841170-99999  2010.0    1.0  15.0  18.0         27.6       20.8   
3  841170-99999  2010.0    1.0  15.0  19.0         28.0       22.0   
4  841170-99999  2010.0    1.0  15.0  20.0         28.0       22.0   

      pressure  wind_direction  wind_speed  sky_condition  precipitation_1h  \
0  1010.787648            30.0         5.1            7.0               NaN   
1  1010.787648            31.0         5.7            7.0               NaN   
2  1011.500000            31.0         5.1            7.0               NaN   
3  1010.787648            32.0         6.2            7.0               NaN   
4  1010.787648            30.0         7.2            7.0               NaN   

   precipitation_6h      STATION NAME CTRY ST CALL  
0               2.1  ELOY ALFARO INTL   EC    SEMT  
1               2.1  ELOY ALFARO INTL   EC    SEMT  
2               2.1  ELOY ALFARO INTL   EC    SEMT  
3               2.1  ELOY ALFARO INTL   EC    SEMT  
4               2.1  ELOY ALFARO INTL   EC    SEMT

In [77]:
weather.loc[weather.station_wban == '841170-99999', 'pressure'].mean()

1010.7876476906303